kava/operations_test.go

```diff
@@ -573,41 +578,339 @@ func TestFeeToOperations(t *testing.T) {
+			for _, op := range ops {
+				if !mustAccAddressFromBech32(op.Account.Address).Equals(st.Account) {
+					continue
+				}
+
+				symbol := op.Amount.Currency.Symbol
+				value, err := types.AmountValue(op.Amount)
+				if value.Sign() != -1 {
+					continue // exit if value is non-negative, as this is not a send
+				}
+				require.NoError(t, err)
+
+				denom, ok := Denoms[symbol]
+				require.True(t, ok)
+
+				opCoins = opCoins.Add(sdk.NewCoin(denom, sdk.NewIntFromBigInt(value.Neg(value))))
+			}
+			assert.True(t, opCoins.IsEqual(st.Balance))
+		}
+	})
+
+	t.Run("coin operation amounts match for recipient", func(t *testing.T) {
+		for _, rt := range receiverTracking {
+			if rt.Account == nil {
+				t.Skip("no recipient")
+			}
+			opCoins := sdk.NewCoins()
+			for _, op := range ops {
+				if !mustAccAddressFromBech32(op.Account.Address).Equals(rt.Account) {
+					continue
+				}
+				symbol := op.Amount.Currency.Symbol
+				value, err := types.AmountValue(op.Amount)
+				if value.Sign() != 1 {
+					continue
+				}
+				require.NoError(t, err)
+
+				denom, ok := Denoms[symbol]
+				require.True(t, ok)
+
+				opCoins = opCoins.Add(sdk.NewCoin(denom, sdk.NewIntFromBigInt(value)))
+			}
+			assert.True(t, opCoins.IsEqual(rt.Balance))
+		}
+
+	})
+
+	t.Run("each sender op has no related ops", func(t *testing.T) {
+
+		for _, op := range ops {
+			value, err := types.AmountValue(op.Amount)
+			require.NoError(t, err)
+			if value.Sign() == -1 {
+				assert.Equal(t, 0, len(op.RelatedOperations))
+			}
+		}
+	})
+
+	t.Run("each recipient op is related to a sender op", func(t *testing.T) {
+
+		for _, op := range ops {
+			value, err := types.AmountValue(op.Amount)
+			require.NoError(t, err)
+			if value.Sign() == 1 {
+				assert.Equal(t, 1, len(op.RelatedOperations))
+				relatedOpIndex := op.RelatedOperations[0].Index
+				relatedOp := ops[relatedOpIndex-ops[0].OperationIdentifier.Index]
+
+				// index matches as expected
+				assert.Equal(t, relatedOpIndex, relatedOp.OperationIdentifier.Index)
+				// values match
+				negatedRelValue, err := types.NegateValue(relatedOp.Amount.Value)
+				require.NoError(t, err)
+				assert.Equal(t, op.Amount.Value, negatedRelValue)
+
+				// currencies match
+				assert.Equal(t, op.Amount.Value, negatedRelValue)
+			}
+		}
+	})
+}
+
+func calculateCoins(ops []*types.Operation) sdk.Coins {
+	coins := sdk.NewCoins()
+	for _, op := range ops {
+		coinAmount, ok := sdk.NewIntFromString(op.Amount.Value)
+		if !ok {
+			panic(fmt.Sprintf("invalid input amount: %s\n", op.Amount.Value))
+		}
+		if coinAmount.IsPositive() {
+			coins = coins.Add(sdk.NewCoin(Denoms[op.Amount.Currency.Symbol], coinAmount))
+		}
+	}
+	return coins
+}
+
+func readABCILogFromFile(file string) sdk.ABCIMessageLogs {
+	txResponse := sdk.TxResponse{}
+	bz, err := ioutil.ReadFile(filepath.Join("mocks", file))
+	if err != nil {
+		panic(err)
+	}
+	cdc := app.MakeCodec()
+	cdc.MustUnmarshalJSON(bz, &txResponse)
+	return txResponse.Logs
```

Comment:
Let's keep our fixtures simple and error if we have more than one log.  Transaction unit tests cover indexing over multiple messages, and operation invariant tests ensure each message correctly starts are the right index.

Let's use pattern above and fail if `len(txreponse.Logs) != 1`.